crates/leon/src/parser.rs

```diff
@@ -0,0 +1,314 @@
+use std::str::FromStr;
+
+use crate::{Item, ParseError};
+
+enum Current {
+    Text(String),
+    Key(String),
+}
+
+impl FromStr for super::Template {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut tokens = Vec::new();
+        let mut current = Current::Text(String::new());
+
+        for (i, c) in s.chars().enumerate() {
+            match (c, &current) {
+                ('{', &Current::Text(ref t)) => {
+                    tokens.push(Item::Text(t.clone()));
+                    current = Current::Key(String::new());
+                }
+                ('{', &Current::Key(ref k)) if k.is_empty() => {
+                    if let Some(Item::Text(mut t)) = tokens.pop() {
+                        t.push('{');
+                        current = Current::Text(t);
+                    } else {
+                        return Err(ParseError {
+                            src: s.to_string(),
+                            unbalanced: Some((i, i + 1).into()),
+                            empty_key: None,
+                        });
+                    }
+                }
+                ('}', &Current::Key(ref k)) => {
+                    tokens.push(Item::Key(k.clone()));
+                    current = Current::Text(String::new());
+                }
+                ('}', &Current::Text(ref k)) if k.chars().last() == Some('}') => {
+                    // skip, that's the escape
+                }
+                (c, &Current::Text(ref t)) => current = Current::Text(format!("{t}{c}")),
+                (c, &Current::Key(ref k)) => current = Current::Key(format!("{k}{c}")),
+            }
+        }
+
+        match current {
+            Current::Text(t) => tokens.push(Item::Text(t)),
+            Current::Key(_) => {
+                return Err(ParseError {
+                    src: s.to_string(),
+                    unbalanced: Some((s.len() - 1, s.len()).into()),
+                    empty_key: None,
+                })
+            }
+        }
+
+        Ok(Self {
+            items: tokens
+                .into_iter()
+                .filter_map(|tok| match tok {
+                    Item::Text(t) if t.is_empty() => None,
+                    Item::Key(k) if k.is_empty() => Some(Err(ParseError {
+                        src: s.to_string(),
+                        unbalanced: None,
+                        empty_key: Some((s.len() - 1, s.len()).into()),
+                    })),
+                    Item::Key(k) => Some(Ok(Item::Key(k.trim().to_string()))),
```

Comment:
You would have to check for `str::is_empty` again after `str::trim`, since `k` could be entirely made of ascii whitespace.